src/hotspot/cpu/x86/stubGenerator_x86_64_dilithium.cpp

```diff
@@ -0,0 +1,1038 @@
+  // be multiplied by the zetas in one set, the rest to another set of vector
+  // registers, then redistribute the addition/substraction results.
+
+  // For levels 0 and 1 the zetas are not different within the 4 xmm registers
+  // that we would use for them, so we use only one, xmm29.
+  loadXmm29(zetas, 0, _masm);
+  __ vpbroadcastd(montQInvModR,
+                  ExternalAddress(dilithiumAvx512ConstsAddr(montQInvModRIdx)),
+                  Assembler::AVX_512bit, scratch); // q^-1 mod 2^32
+  __ vpbroadcastd(dilithium_q,
+                  ExternalAddress(dilithiumAvx512ConstsAddr(dilithium_qIdx)),
+                  Assembler::AVX_512bit, scratch); // q
+
+  // load all coefficients into the vector registers Zmm_0-Zmm_15,
+  // 16 coefficients into each
+  load4Xmms(xmm0_3, coeffs, 0, _masm);
+  load4Xmms(xmm4_7, coeffs, 4 * XMMBYTES, _masm);
+  load4Xmms(xmm8_11, coeffs, 8 * XMMBYTES, _masm);
+  load4Xmms(xmm12_15, coeffs, 12 * XMMBYTES, _masm);
+
+  // level 0 and 1 can be done entirely in registers as the zetas on these
+  // levels are the same for all the montmuls that we can do in parallel
+
+  // level 0
+  montMul64(xmm16_19, xmm8_11, xmm29_29, xmm16_27, _masm);
+  sub_add(xmm8_11, xmm0_3, xmm0_3, xmm16_19, _masm);
+  montMul64(xmm16_19, xmm12_15, xmm29_29, xmm16_27, _masm);
+  loadXmm29(zetas, 512, _masm); // for level 1
+  sub_add(xmm12_15, xmm4_7, xmm4_7, xmm16_19, _masm);
+
+  // level 1
+
+  montMul64(xmm16_19, xmm4_7, xmm29_29, xmm16_27, _masm);
+  loadXmm29(zetas, 768, _masm);
+  sub_add(xmm4_7, xmm0_3, xmm0_3, xmm16_19, _masm);
+  montMul64(xmm16_19, xmm12_15, xmm29_29, xmm16_27, _masm);
+  sub_add(xmm12_15, xmm8_11, xmm8_11, xmm16_19, _masm);
+
+  // levels 2 to 7 are done in 2 batches, by first saving half of the coefficients
+  // from level 1 into memory, doing all the level 2 to level 7 computations
+  // on the remaining half in the vector registers, saving the result to
+  // memory after level 7, then loading back the coefficients that we saved after
+  // level 1 and do the same computation with those
+
+  store4Xmms(coeffs, 0, xmm0_3, _masm);
+  store4Xmms(coeffs, 4 * XMMBYTES, xmm4_7, _masm);
+  store4Xmms(coeffs, 8 * XMMBYTES, xmm8_11, _masm);
+  store4Xmms(coeffs, 12 * XMMBYTES, xmm12_15, _masm);
+
+  __ movl(iterations, 2);
+
+  __ align(OptoLoopAlignment);
+  __ BIND(L_loop);
+
+  __ subl(iterations, 1);
+
+  // level 2
+  load4Xmms(xmm12_15, zetas, 2 * 512, _masm);
+  montMul64(xmm16_19, xmm2367, xmm12_15, xmm16_27, _masm);
+  load4Xmms(xmm12_15, zetas, 3 * 512, _masm); // for level 3
+  sub_add(xmm2367, xmm0145, xmm0145, xmm16_19, _masm);
+
+  // level 3
+
+  montMul64(xmm16_19, xmm1357, xmm12_15, xmm16_27, _masm);
+  sub_add(xmm1357, xmm0246, xmm0246, xmm16_19, _masm);
+
+  // level 4
+  loadPerm(xmm16_19, perms, nttL4PermsIdx, _masm);
+  loadPerm(xmm12_15, perms, nttL4PermsIdx + 64, _masm);
+  load4Xmms(xmm24_27, zetas, 4 * 512, _masm); // for level 3
```

Comment:
The comment // for level3 is not relevant here and could be removed.

Reply:
Ooops. Deleted the comment.